### ISSUE TITLE
False "incompatible default" error for `def f(v: T = 0) -> T`, `T = TypeVar("T", Literal[0], Literal[1])`

### ISSUE THREAD

**Issue by tyralla** [python/mypy#15381]
**Bug Report**

False positive when assigning a default value to a function argument that is annotated with a type variable with literal value restrictions.

**To Reproduce**

```python
from typing import Literal, TypeVar

T = TypeVar("T", Literal[0], Literal[1])

def f(v: T = 0) -> T: ...  # E: Incompatible default for argument "v" (default has type "Literal[0]", argument has type "Literal[1]")

reveal_type(f())   # N: Revealed type is "Literal[0]
reveal_type(f(0))  # N: Revealed type is "Literal[0]
reveal_type(f(1))  # N: Revealed type is "Literal[1]
reveal_type(f(2))  # E: error: Value of type variable "T" of "f" cannot be "Literal[2]"
                   # N: Revealed type is "Literal[2]
```

**Expected Behavior**

No error report in the `def f(v: T = 0) -> T: ...` line.

**Your Environment**

- Mypy version used: 1.3.0
- Python version used: 3.11

**Comment by hauntsaninja** [comment]
Duplicate of https://github.com/python/mypy/issues/3737
